### ISSUE TITLE
[PowerShell@2]: Task reports success if PowerShell.exe is killed externally

### ISSUE THREAD

**Issue by PST2** [microsoft/azure-pipelines-tasks#15239]
## Required Information

**Question, Bug, or Feature?**  
*Type*: Bug/Feature

**Enter Task Name**: Powershell@2

## Environment
- Server - Azure Pipelines

- Agent - Private; v2.166.3 agent on Windows Server Core 10.0.19042

## Issue Description

Background: our org has mandated a 3rd party antimalware product on all boxes & VMs, including our onprem build agents. Part of its policy enforcement targets certain uses of PowerShell; if it sees something it doesn't like, it kills the PowerShell process. Our dev team is not in a position to change the product usage or the antimalware policies, which are pretty broad and trap an undocumented list of (sometimes useful) functionality.

Where this affects the PowerShell@2 task is that if a script invokes something on the blacklist, the killing of the PowerShell process is detected by the agent as successful completion of the script step. This means we remain unaware that script actions were silently not performed, making any subsequent issues in our software take longer to investigate (particularly if this happens on a deployment environment rather than a build agent).

We could add Write-Host logging throughout our scripts to track whether they completed properly or not, but this would become a manual task to inspect rather than seeing the failure light up immediately. Given that the task first writes the PowerShell script to a temporary file before execution, perhaps it could add something (eg. a temporary lock file created before the ps1 executes and deleted by the ps1 as the final step) to detect whether PowerShell.exe actually fully executed the script vs "abnormally" terminating?

### Task logs

Added a Write-Host before the suspected problem lines and another at the end of our script to announce completion. Only the first reached the task log.
[Ps-killed-log.zip](https://github.com/microsoft/azure-pipelines-tasks/files/7089423/Ps-killed-log.zip)

### Error logs

Issue is that no error is logged and task reports success. Debug log shows externally killed PowerShell process returned exit code 0.

**Comment by EzzhevNikita** [comment]
Hello @PST2,
Thank you for logging the ticket.
Unfortunately, I can't repro this issue. As I see the killing of processes started by task PowerShell@2 causes returning of exit code '-1' and failure of the task.
Could you please share YAML of the affected pipeline or a YAML with a minimal reproducible sample to allow me repro this on my side?

**Comment by PST2** [comment]
I suspect this one might be difficult to locally rep without having the same (3rd party) antimalware installed. This is enough to do it on our onprem agents:
```yaml
- task: PowerShell@2
  displayName: Should Not Pass But Does
  inputs:
    targetType: inline
    script: |
      Write-Host 'Begin step...'
      $conn = new-object system.data.sqlclient.sqlconnection('server=<snip>;Database=<snip>;Integrated Security=true')
      $conn.Open()
      Write-Host '...End step.'
```
As the end of the diagnostics-enabled log shows, we never make it to the second Write-Host, PowerShell.exe reports exit code 0, and the step gets green-ticked despite not executing everything sent to it:
```
========================== Starting Command Output ===========================
##[debug]Entering Invoke-VstsTool.
##[debug] Arguments: '-NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'C:\<snip>\_work\_temp\668a788b-f63e-4870-9dbd-06121f5bf1e4.ps1'"'
##[debug] FileName: 'C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe'
##[debug] WorkingDirectory: 'C:\<snip>\_work\1\s'
"C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'C:\<snip>\_work\_temp\668a788b-f63e-4870-9dbd-06121f5bf1e4.ps1'"
Begin step...
##[debug]Exit code: 0
##[debug]Leaving Invoke-VstsTool.
##[debug]Leaving C:\<snip>\_work\_tasks\PowerShell_e213ff0f-5d5c-4791-802d-52ea3e7be1f1\2.190.0\powershell.ps1.
Finishing: Should Not Pass But Does
```

A slightly different way of illustrating what's going on here is replicating the behaviour by hand in a CMD session. The antimalware sees the attempt to make a DB connection from within PowerShell and somehow instantly kicks us back to the CMD prompt with a zero exit code for the PowerShell process:
```
D:\>echo %errorlevel%
0

D:\>powershell
Windows PowerShell
Copyright (C) Microsoft Corporation. All rights reserved.

Try the new cross-platform PowerShell https://aka.ms/pscore6

PS D:\> $conn = new-object system.data.sqlclient.sqlconnection('server=<snip>;database=<snip>;Integrated Security=true;')
PS D:\> $conn.open()

D:\>echo %errorlevel%
0

D:\>
```

**Comment by EzzhevNikita** [comment]
@PST2 Thank you for providing me these details. Please provide the PowerShell version installed on used agent also. It can be got by command `(Get-Host).version`

**Comment by PST2** [comment]
Sure:
```
PS C:\> (Get-Host).Version

Major  Minor  Build  Revision
-----  -----  -----  --------
5      1      19041  1237
```

**Comment by EzzhevNikita** [comment]
@PST2, thank you for providing the info.
Unfortunately, we are still not able to reproduce your issue. 
Could you please confirm that CMD output above was got from the same agent? Did you see the same issue on other machines/agents?
How frequently has this issue occurs? Did you see any correlations between reproes of the issue and any additional factors (network quality, agent load, etc.)?

**Comment by PST2** [comment]
I don't believe this has anything to do with the agent per se, or the conditions (other than antimalware) under which it is running.

I have replicated the CMD behaviour on half a dozen machines on our domain, all of which have the same antimalware but are different OS versions (Win10, WS2019, WSC20H1), both with and without the agent installed. The described behaviour is not intermittent and happens 100% of the time. Moreover, on digging into this further, messages like this appear in event viewer in all cases (vendor/product name snipped as I've been advised not to post them publicly):

> Information: The application C:\windows\system32\windowspowershell\v1.0\powershell.exe attempted to establish a TCP/1433 connection to (DB IP):1433 from (local IP):59948. The operation was blocked and the application terminated by {antimalware product name}.

SysInternals Process Monitor agrees on the exit code being 0:
![image](https://user-images.githubusercontent.com/42988215/135023471-a9983c31-7866-4dc1-bc5d-64efc6475716.png)

I appreciate that this is not an easy scenario to, say, write a test for. It looks like the task uses a wrapper script to execute the user-supplied one. Perhaps before PowerShell.exe is launched a temp file could be created, with the last action of the wrapper script being to delete it -- and that file continuing to exist after PowerShell.exe returns being an indicator of abnormal termination?

**Comment by EzzhevNikita** [comment]
@PST2, thank you for providing info.
As I understand from all info about issue, it's not related to PowerShell task, it's related to compatibility issue for PowerShell shell tool and used antimalware suite. And adding of check for finish of PowerShell script won't solve this issue. All scripts with activities like described in the issue will fail after this possible fix (in accordance with info that this behavior happens 100% of the time).
So I would suggest to check configuration of used antimalware suite to resolve root cause of the issue. Also I think that you can ask support team of used antimalware suite about compatibility with PowerShell tool, it's strange behavior to block PowerShell script but allow this to finish with exit code 0.

**Comment by PST2** [comment]
Yes, I filed this issue after taking the matter up with IT on our side. Altering the blocking behaviour is a non-starter -- all requests to do so get rejected -- hence this request to detect such an occurrence.

Just to reiterate, this issue is not about the fact these steps get blocked; I agree that is outside the scope of the Task's ability to fix. Rather, it's that the step doesn't result in a failure when the script isn't executed all the way through due to the process being terminated in this manner -- in our opinion, this should fail the build job (as it would if one of the lines in the script itself threw an exception, or left lastexitcode nonzero at the end).

It would be enough for us to just get a ❌ and error message saying something like "Abnormal script termination" from the step when this happens.

A similar workaround would be for us to use a logging command at the end to set a pipeline variable and check the result in a subsequent step, but as with other detection workarounds we'd have to remember to add the required YAML/PS every time (with the resulting increase in job step counts as well).

**Comment by EzzhevNikita** [comment]
@PST2 , root cause of this issue is strange behavior of PowerShell tool after blocking this by antimalware tool. I checked many possible cases with blocking or killing of PowerShell process by Stop-Process command, "kill process" feature in Task Manager, Windows Defender and Firewall tools. But each of the cases finished with returning of non-zero exit code. And all these cases are successfully handled by PowerShell task. I didn't find any case with killing PowerShell process and returing exit code `0`.
So it's too hard to repro this issue correctly and it's better to fix root cause of the issue.

**Comment by PST2** [comment]
Awaiting vendor response.
